internal/encoding/asn1/asn1.go

```diff
@@ -0,0 +1,219 @@
+		}, nil
+	}
+	// constructed value
+	rootConstructed := constructedValue{
+		identifier:  identifier,
+		expectedLen: contentLen,
+	}
+
+	// start depth-first decoding with stack
+	valueStack := []*constructedValue{&rootConstructed}
+	for len(valueStack) > 0 {
+		stackLen := len(valueStack)
+		// top
+		v := valueStack[stackLen-1]
+
+		if v.expectedLen < 0 {
+			return nil, ErrInvalidBerData
+		}
+
+		if v.expectedLen == 0 {
+			// calculate the length of the constructed value
+			for _, m := range v.members {
+				v.length += m.EncodedLen()
+			}
+
+			// pop the constructued value
+			valueStack = valueStack[:stackLen-1]
+			break
+		}
+
+		for v.expectedLen > 0 {
+			identifier, contentLen, berValueLen, isPrimitive, r, err = decodeMetadata(r)
+			if err != nil {
+				return nil, err
+			}
+			if isPrimitive {
+				// primitive value
+				pv := primitiveValue{
+					identifier: identifier,
+					content:    r[:contentLen],
+				}
+				r = r[contentLen:]
+				v.expectedLen -= berValueLen
+				v.members = append(v.members, &pv)
+			} else {
+				// constructed value
+				cv := constructedValue{
+					identifier:  identifier,
+					expectedLen: contentLen,
+				}
+				v.expectedLen -= berValueLen
+				v.members = append(v.members, &cv)
+				valueStack = append(valueStack, &cv)
+				// break to start decoding the new constructed value
+				break
+			}
+		}
+	}
+	if len(r) > 0 {
+		return nil, ErrBytesAtTheEnd
+	}
+	return rootConstructed, nil
+}
+
+func decodeMetadata(r []byte) ([]byte, int, int, bool, []byte, error) {
+	length := len(r)
+	identifier, r, err := decodeIdentifier(r)
+	if err != nil {
+		return nil, 0, 0, false, nil, err
+	}
+	contentLen, r, err := decodeLen(r)
+	if err != nil {
+		return nil, 0, 0, false, nil, err
+	}
+
+	if contentLen > len(r) {
+		return nil, 0, 0, false, nil, ErrEarlyEOF
+	}
+	isPrimitive := identifier[0]&0x20 == 0
+	metadataLen := length - len(r)
+	berValueLen := metadataLen + contentLen
+	return identifier, contentLen, berValueLen, isPrimitive, r, nil
+}
+
+// decodeIdentifier decodes decodeIdentifier octets.
+func decodeIdentifier(r []byte) ([]byte, []byte, error) {
+	offset := 0
+	if len(r) < 1 {
+		return nil, nil, ErrEarlyEOF
+	}
+	b := r[offset]
+	offset++
+
+	// high-tag-number form
+	if b&0x1f == 0x1f {
+		for offset < len(r) && r[offset]&0x80 == 0x80 {
+			offset++
+		}
+	}
+	return r[:offset], r[offset:], nil
+}
+
+// decodeLen decodes length octets.
+// Indefinite length is not supported
+func decodeLen(r []byte) (int, []byte, error) {
+	offset := 0
+	if len(r) < 1 {
+		return 0, nil, ErrEarlyEOF
+	}
+	b := r[offset]
+	offset++
+	switch {
+	case b < 0x80:
+		// short form
+		return int(b), r[offset:], nil
+	case b == 0x80:
+		// Indefinite-length method is not supported.
+		return 0, nil, ErrUnsupportedIndefinedLen
+	}
+
+	// long form
+	n := int(b & 0x7f)
+	if n > 4 {
+		// length must fit the memory space of the int type.
+		return 0, nil, ErrUnsupportedLen
+	}
+	var length int
```

Comment:
it is better to have `uint64` as we are doing bit operation
```suggestion
	var length uint64
```

Reply:
Updated.